benchmarks/cpp/src/qsort.cpp

```diff
@@ -0,0 +1,79 @@
+#include <algorithm>
+#include <array>
+#include <asio.hpp>
+#include <chrono>
+#include <iostream>
+#include <random>
+#include <vector>
+
+using namespace asio;
+using namespace std;
+using namespace std::chrono;
+
+void insertionSort(vector<int>& arr);
+int partition(vector<int>& arr);
+
+void shuffle(vector<int>& arr) {
+    mt19937 rng(random_device{}());
+    shuffle(begin(arr), end(arr), rng);
+}
+
+void quickSort(vector<int>& arr) {
+    if (arr.size() <= 32) {
+        insertionSort(arr);
+    } else {
+        io_context ctx;
+        steady_timer timer{ctx};
+        vector<int> arr1, arr2;
```

Comment:
These seem like they could remove by having quickSort() take in a start and end along with `arr`. Avoids heap allocating 2 intermediate vectors and copying them back at the end

Reply:
Iterator's... then need +1 param (begin and end) on quicksort and partitions. (maybe!!)

Reply:
I tried to replace it with std::array, but I got segfault during execution!

Reply:
I dont mean to replace the intermediary vectors with arrays, but to remove them in favor of slices:
```
qsort(vec, start, end):
  if (end - start) <= 32:
    return isort(vec, start, end)
  pivot = partition(vec, start, end)
  dispatch(qsort, vec, start, start + pivot)
  dispatch(qsort, vec, start + pivot, end)
  run()
```

Reply:
I have made the suggested changes, but an unexpected error occurs in running step.
```c++
void quickSort(std::vector<int>& arr, int start, int end) {
    if (end - start <= 32) {
        insertionSort(arr, start, end); //custom insertionSort changed too
    } else {
        asio::io_context ctx;
        asio::steady_timer timer{ctx};

        const auto pivot = partition(arr, start, end); // custom partition changed too
        
        // Create a strand to wrap the async calls to quickSort
        asio::io_context::strand strand{ctx};
        auto quickSortWrapper = [&](auto&& arr, auto start, auto end) {
            quickSort(arr, start, end);
        };
        
        // Use the strand to dispatch the async calls to quickSort
        strand.dispatch(std::bind(quickSortWrapper, std::ref(arr), start, start + pivot));
        strand.dispatch(std::bind(quickSortWrapper, std::ref(arr), start + pivot, end));
        
        // Run the io_context to process the dispatched tasks
        ctx.run();
    }
}
```
**output:**
```bash
filling
shuffling
running
The following command terminated unexpectedly:
```

Reply:
try printing out the start & end of each invocation. It may not be handling negative sizes (end > start)

Reply:
**output**

```bash
filling
shuffling
running
[...] // skip to final list
14 - 740
14 - 753
14 - 766
14 - 779
14 - 792
14 - 805
14 - 685
14 - 698
The following command terminated unexpectedly:
```

Reply:
I will commit the change!